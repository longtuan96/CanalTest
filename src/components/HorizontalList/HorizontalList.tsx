import { ReactNode, useEffect } from "react";
import { ArrowLeftIcon, ArrowRightIcon } from "@heroicons/react/24/solid";
import { AxiosError } from "axios";
import useScrollControl from "../../hooks/useScrollControl";

export type HorizontalListProps = {
  title: string;
  children?: ReactNode;
  error?: AxiosError | unknown;
};

export default function HorizontalList(props: HorizontalListProps) {
  const { title, children, error } = props;
  const {
    ref: scrollDivRef,
    isAtStart,
    isAtEnd,
    scroll,
    checkScroll,
  } = useScrollControl();

  useEffect(() => {
    checkScroll();
  }, [checkScroll, children]);

  return (
    <div className="flex flex-col gap-2 relative">
      <h4 className="text-2xl font-semibold text-white mx-6 sm:mx-8 md:mx-12">
        {title}
      </h4>
      {!isAtStart && (
        <button className="absolute bg-black bg-opacity-50 shadow-md backdrop-blur p-2 z-10 rounded left-2 top-1/2 -translate-y-1/2">
          <ArrowLeftIcon
            className="w-6 h-6"
            title="scroll left"
            onClick={scroll.bind(null, true)}
          />
        </button>
      )}
      {!isAtEnd && (
        <button className="absolute bg-black bg-opacity-50 shadow-md backdrop-blur p-2 z-10 rounded right-2 top-1/2 -translate-y-1/2">
          <ArrowRightIcon
            className="w-6 h-6"
            title="scroll right"
            onClick={scroll.bind(null, false)}
          />
        </button>
      )}
      <div
        ref={scrollDivRef}
        className="flex flex-row gap-6 overflow-x-auto py-2 px-6 sm:px-8 md:px-12 relative scrollbar-hide"
      >
        {!!error && <span>{"Fetch Data failed!"}</span>}
        {children}
      </div>
    </div>
  );
}
