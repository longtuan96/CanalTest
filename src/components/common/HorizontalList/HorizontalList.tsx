import { ReactNode, useEffect } from "react";
import {
  ChevronDoubleLeftIcon,
  ChevronDoubleRightIcon,
} from "@heroicons/react/24/solid";
import { AxiosError } from "axios";
import useScrollControl from "../../../hooks/useScrollControl";

export type HorizontalListProps = {
  title: string;
  children?: ReactNode;
  error?: AxiosError | unknown;
};

export default function HorizontalList(props: HorizontalListProps) {
  const { title, children, error } = props;
  const {
    ref: scrollDivRef,
    isAtStart,
    isAtEnd,
    scroll,
    checkScroll,
  } = useScrollControl();

  useEffect(() => {
    checkScroll();
  }, [checkScroll, children]);

  return (
    <div className="relative flex flex-col gap-2">
      <h4 className="mx-6 text-2xl font-semibold text-white sm:mx-8 md:mx-12">
        {title}
      </h4>
      {!isAtStart && (
        <button className="absolute left-2 top-1/2 z-10 -translate-y-1/2 rounded-full bg-black bg-opacity-50 p-2 shadow-md backdrop-blur">
          <ChevronDoubleLeftIcon
            className="h-6 w-6"
            title="scroll left"
            onClick={scroll.bind(null, true)}
          />
        </button>
      )}
      {!isAtEnd && (
        <button className="absolute right-2 top-1/2 z-10 -translate-y-1/2 rounded-full bg-black bg-opacity-50 p-2 shadow-md backdrop-blur">
          <ChevronDoubleRightIcon
            className="h-6 w-6"
            title="scroll right"
            onClick={scroll.bind(null, false)}
          />
        </button>
      )}
      <div
        ref={scrollDivRef}
        className="scrollbar-hide relative flex flex-row gap-6 overflow-x-auto px-6 py-2 sm:px-8 md:px-12"
      >
        {!!error && <span>{"Fetch Data failed!"}</span>}
        {children}
      </div>
    </div>
  );
}
