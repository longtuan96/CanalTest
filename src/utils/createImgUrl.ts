import MoviePlaceHolder from "../assets/landscape-placeholder-svgrepo-com.svg";
import PeoplePlaceHolder from "../assets/person-svgrepo-com.svg";

export default function createImgUrl(
  type: "people" | "movie",
  path?: string | null,
  size = "original",
) {
  if (!path) return type === "people" ? PeoplePlaceHolder : MoviePlaceHolder;
  return `${import.meta.env.VITE_THEMOVIEDB_IMG_URL}/${size}${path}`;
}
