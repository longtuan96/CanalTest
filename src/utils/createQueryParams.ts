export default (
  queryParams: Record<string, string | number | null | undefined> = {},
) => ({
  ...queryParams,
  api_key: import.meta.env.VITE_THEMOVIEDB_KEY,
});
