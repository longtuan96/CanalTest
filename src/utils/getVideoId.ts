import { VideoItem } from "../services/video/video.type";

export default function getVideoId(data: VideoItem[] | undefined) {
  if (data) {
    return data.find((item) => item.name === "Official Trailer")?.key;
  } else return undefined;
}
