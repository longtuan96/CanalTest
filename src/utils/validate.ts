import { ZodObject, ZodRawShape } from "zod";

export default function <T>(schema: ZodObject<ZodRawShape>, data: T) {
  const parseResult = schema.safeParse(data);
  if (!parseResult.success) {
    throw Error("Validation Failed!");
  }
  return parseResult.data as T;
}
