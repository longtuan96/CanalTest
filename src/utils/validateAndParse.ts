import { ZodObject, ZodRawShape } from "zod";

export default function validateAndParse<T>(
  schema: ZodObject<ZodRawShape>,
  data: T,
) {
  const parseResult = schema.safeParse(data);
  if (!parseResult.success) {
    throw Error("Validation Failed!");
  }
  return parseResult.data as T;
}
