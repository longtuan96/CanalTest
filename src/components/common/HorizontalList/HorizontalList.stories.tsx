import type { Meta, StoryObj } from "@storybook/react";

import HorizontalList from "./HorizontalList";
import { moviePopularMockData } from "../../../services/movie/movie.mock";
import { MovieCard } from "../../MovieCard/MovieCard";
import { creditMockData } from "../../../services/credit/credit.mock";
import { PeopleCard } from "../../PeopleCard/PeopleCard";

const meta = {
  component: HorizontalList,
} satisfies Meta<typeof HorizontalList>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Movie: Story = {
  args: {
    title: "Movie",
    children: moviePopularMockData.results.map((item) => (
      <MovieCard data={item} key={item.id} />
    )),
  },
};

export const People: Story = {
  args: {
    title: "Cast",
    children: creditMockData.cast.map((item) => (
      <PeopleCard data={item} key={item.id} />
    )),
  },
};

export const Error: Story = {
  args: {
    title: "Error",
    error: true,
  },
};
