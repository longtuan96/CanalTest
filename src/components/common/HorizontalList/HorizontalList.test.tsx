import { describe, expect, test } from "vitest";
import HorizontalList from "./HorizontalList";
import { getTree, render, screen } from "../../../test/utils";

const title = "test";
const children = "Something";

describe("HorizontalList", () => {
  test("render correctly", () => {
    expect(
      getTree(<HorizontalList title={title}>{children}</HorizontalList>),
    ).toMatchSnapshot();
  });
  test("title is a heading", () => {
    render(<HorizontalList title={title}>{children}</HorizontalList>);
    screen.getByRole("heading", { name: title });
  });
});
