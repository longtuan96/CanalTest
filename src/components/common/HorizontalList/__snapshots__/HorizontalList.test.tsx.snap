// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HorizontalList > render correctly 1`] = `
<div
  className="relative flex flex-col gap-2"
>
  <h4
    className="mx-6 text-2xl font-semibold text-white sm:mx-8 md:mx-12"
  >
    test
  </h4>
  <button
    className="absolute right-2 top-1/2 z-10 -translate-y-1/2 rounded bg-black bg-opacity-50 p-2 shadow-md backdrop-blur"
  >
    <svg
      aria-hidden="true"
      className="h-6 w-6"
      data-slot="icon"
      fill="currentColor"
      onClick={[Function]}
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        scroll right
      </title>
      <path
        clipRule="evenodd"
        d="M12.97 3.97a.75.75 0 0 1 1.06 0l7.5 7.5a.75.75 0 0 1 0 1.06l-7.5 7.5a.75.75 0 1 1-1.06-1.06l6.22-6.22H3a.75.75 0 0 1 0-1.5h16.19l-6.22-6.22a.75.75 0 0 1 0-1.06Z"
        fillRule="evenodd"
      />
    </svg>
  </button>
  <div
    className="relative flex flex-row gap-6 overflow-x-auto px-6 py-2 sm:px-8 md:px-12"
  >
    Something
  </div>
</div>
`;
